README.md
# water-meter-system-complete

This repository describes a readout of an analog water meter with the help of a camera and image processing, including neural network processing to extract the values.
The interface is a local HTTP-server, that takes an image as input, processes it and gives as an output the readout of the water meter counter, including the subdigits.

Although the server can be installed manually in a windows as well as an Linux system, it is strongly recommended to use the provided docker containers. The installation is due to the use of image processing (OpenCV, ...) and Neural Network Systems (Tensorflow) not really straight forward and I needed a lot of try and error. 

**Therefore I have decided to capsule the server in a docker container**. 

This is available for an Intel based System (e.g. Synology Docker) as well as an Raspberry PI Version.


## Docker-Versions
| Label | 	Content  | Comments   |
| -------------- | ------------- | ------------- |
| rolling | Experimental version for amd64-systems (e.g. Intel processors) | newest features, not fully tested |
| raspi-rolling | Experimental version for armv7-systems (Raspberry PI3 and higher) | newest features, not fully tested  |
| latest | Latest stable version for amd64-systems (e.g. Intel processors) | Currently identical to 6.1.1 (2020-04-23) |
| raspi-latest | Latest stable version for armv7-systems (Raspberry PI3 and higher) | Currently identical to 6.1.1 (2020-04-23) |
| v5.x | Persistant prevalue, modified docker structure | Details see below |
| v4.x | Update to Tensorflow 2.0, fully automated build | Details see below |
| v3.x | Tensorflow 1.4, manual build | Details see below  |

## Running docker
Choose for the fitting docker tag and run the server with the following parameters:

```sudo docker run -p 3000:3000 --mount type=bind,source=/PATH_TO_LOCAL_CONFIG,target=/config --mount type=bind,source=/PATH_TO_LOCAL_LOG,target=/log jomjol/wasserzaehler:DOCKER_TAG```

#### Paramters
| Parameter | 	Meaning  | Example |
| -------------- | ------------- | ------------- |
| PATH_TO_LOCAL_CONFIG | Configuration parameters are stored in this path - easeast way to handle is a local copy of this directory | ```/home/pi/config``` |
| PATH_TO_LOCAL_LOG | Logging paramters and images are stored in in this path - easeast way to handle is a local copy of this directory | ```/home/pi/config``` |
| DOCKER_TAG | Docker tag for the correct docker version | ```raspi-latest``` |

The config and the log directory can be empty at the very first start. They will be loaded with a default configuratio, that can be modified afterwards.

## Changelog - lastest version

##### 7.1.2 (2020-05-29)
* Updated CNN for digital digits (v6.1.1)
##### 7.1.1 (2020-05-24)
* Error-Correction in config.ini (h5 --> tflite)
##### 7.1.0 (2020-05-14)
* ErrorMessage in case of wrong CNN-files (old h5, instead of supported tflite)
* Neural Network Files updated to newest version - trained with new user images (v6.1.0)
##### 7.0.1 (2020-04-27)
* Error-Correction: saving jpg-logs 
##### 7.0.0 (2020-02-20) **ATTENTION: cnn-files not downward compatible **
* Major Update: use tflite runtime instead of full Tensorflow - **ATTENTION: cnn-files not downward compatible **
* Internal change of CNN-Usage library structure
### **Attention: 6.1.1 is the last version, with h5-files support. Future versions will only support h5-files!!!**
##### 6.1.1 (2020-04-23)
* Last Version with h5-File-Support

### [Full Changelog](https://github.com/jomjol/water-meter-system-complete/blob/raspi-rolling/Changelog.md)


The overall system with description of the single steps is described here: [https://github.com/jomjol/water-meter-measurement-system](https://github.com/jomjol/water-meter-measurement-system)

A graphical overview about the steps is shown in the following flow:

<img src="https://github.com/jomjol/water-meter-system-complete/blob/master/images/signal_flow.png" width="800"> 

## Setup

To run the Python code copy the whole [code](code) directory including subdirectory.

Path are relative, so it should run immediately with the following command:
* `pip install requirements.txt`
* `python wasserzaehler.py`

### Configuration

The configuration is stored in the subdirectory `config`. In the Ini-file the CNN-Network to be loaded is listed. Configuration of the neural network (*.h5) itself is stored in the subdirectory `neuralnets`.
Detailed information on config.ini see [Config_Description.md](Config_Description.md)

##### Consistency Check
There is a consistency check of the readout value implemented. Prequesite for this check is a storage of the last full readout (without "N"), which can be achieved by the parameter "usePreValue".

		
	
## Running the server

The server is listening to port 3000 and accepts requests in the following syntac:

* http://server-ip:3000/wasserzaehler.html

| Parameter | Meaning | example |
| --------- | ------- | ------- |
| server-ip | address of the node-server running the script | `localhost` |

Without any parameter and correct setting in the CONFIG.INI the server responses with an readout of the water meter:
<img src="https://github.com/jomjol/water-meter-system-complete/blob/master/images/server_output.png" width="400">

The output of the server are 3 numbers, separated by a tabulator.

| Number | Meaning | 
| --------- | ------- |
| First number | Full readout, including main digits and subdigits, leading zeros are suppresed |
| Second number | Direct readout of the digital digits, including leading zeros |
| Third number | post digit numbers |

##### Remark
If a digit cannot be recognized, e.g. because it is half between 2 digits, then instead of the number a "N" is written at the corresponding position. In this case a direct conversion to a number will not work. Additional information (e.g. last valid full reading) needs to be used to extrapolate the missing digit (see below).


### Optional parameters

* http://server-ip:3000/wasserzaehler.html?url=http://picture-server/image.jpg&full
* http://server-ip:3000/wasserzaehler.html?usePreValue
* http://server-ip:3000/wasserzaehler.html?single

| Parameter | Meaning | example |
| ----- | ------- | ------ |
| url | url to a dedicated picture to be analysed | `url=http://picture-server/image.jpg` |
| full | response extended by details on readout process | `full` |
| usePreValue | if available the last fully valid readout is used to complete unambigoius digits ('N'). The prevalue can be set manuelly by 'setPreValue.html' - see below | `usePreValue` |
| single | only a singel number is given back instead of combinded value, digital and analog readouot | `single` |

The paramaters can be combined arbitrary.

Example with parameter `full`:

<img src="https://github.com/jomjol/water-meter-system-complete/blob/master/images/sever_output_full.png" width="400">

## Additional Settings
* http://server-ip:3000/roi.html

The page `roi.html` return the image including the ROIs visible. This is usefull to check for correct setting:
<img src="https://github.com/jomjol/water-meter-system-complete/blob/master/images/roi_masked.jpg" width="400">

* http://server-ip:3000/setPreValue.html?value=401.57

The page `setPreValue.html` stores the number given in the parameter `value` to initiate the internal storage of a valid data. This can be used, in case the digital counter is between two number and cannot be readout uniquely at the moment.

| Parameter | Meaning | example |
| --------- | ------- | ------- |
| value | valid setting of water meter readout | `value=401.57` |

* http://server-ip:3000/version.html

The page `version.html` returns a version number.
   





Changelog.md
## Changelog
##### 6.1.1 (2020-04-23)
* Update CNN-Files to v6.0.0, v6.0.1
##### 6.1.0 (2020-04-17)
* MinImageSize configuration parameters added. This will detect damaged images. (Extension from Zwer2k)
##### 6.0.0 (2020-03-16)
* Update Tensorflow to v2.1 - downgrade Python to v3.6 (to enable compiling for Synology (w/o AVX2))
##### 5.6.1 (2020-03-12)
* Correct error in docker commmand (remove space character in bind)
##### 5.6.0 (2020-03-07)
* Internal update of config.ini handling (started)
* Update roi.html
##### 5.5.2 (2020-02-23)
* Modification of Errortext in case RateToHigh, NegativeRate (Blank after "Error", e.g. "ErrorRateToHigh" --> "Error - RateToHigh")
* Change MaxRateValue from 0.1 to 0.2 in Default Config.ini
##### 5.5.1 (2020-02-15)
* Update CNN-Digital to v5.0.0 (improved training data from iobroker users)
* Update CNN-Analog to v5.0.0 (improved training data from iobroker users)
##### 5.5.0 (2020-02-12)
* Update CNN-Digital to v4.2.0 (improved training data from iobroker users)
##### 5.4.9 (2020-01-27)
* Extensdion to json output: ../wasserzaehler.json
##### 5.3.0 (2020-01-08)
* Integration of storage of prevalue in file to reload on startup
* Correction of drawing analog counter ROIs even if they are disabled
##### 5.2.0 (2020-01-03)
* Raspberry Version: Remove autorestart (not working) - instead: use cron job for regular restart to handle tensorflow memory leak [Setting up cron job](https://github.com/jomjol/water-meter-system-complete/blob/raspi-rolling/Raspi-Cron-Job.md)
##### 5.1.0 (2019-12-28)
* Raspberry Version: Autorestart on Python Crash of wasserzaehler.py
##### 5.0.0 (2019-12-28)
* Separate environmental setup to dedicated Docker images (for Raspberry: raspi-opencv-tensorflow and for Synology (Intel w/o AVX2): synology-opencv-tensorflow)
##### 4.4.0 (2019-12-27)
* Change to new base image: export OpenCV and Tensorflow in new base images (for Raspberry and Synology Version)
##### 4.3.0 (2019-12-20)
* Update roi.jpg on every run
##### 4.2.1 (2019-12-17)
* Workaround for memory leak in tensorflow function predict() - see [https://github.com/keras-team/keras/issues/13118](https://github.com/keras-team/keras/issues/13118)
##### 4.2.0 (2019-12-07)
* Reading of analog counters is enabled or disabled via config.ini **special thanks to alikanarya**
##### 4.1.1 (2019-11-29)
* Error correction in ReadDigitalDigitClass.py
##### 4.1.0 (2019-11-25)
* Upgrade to Tensorflow 2.0
* Changed image processing within CNN to Pillow (instead of OpenCV)
* Training analog and digital CNN with additional type of counter and digits respectivelely
##### 3.0.2 (2019-11-07)
* Correct error for disable Logging
##### 3.0.1 (2019-11-03)
* Update to Pillow 6.2.0
##### 3.0.0 (2019-10-06)
* Impementation of optional consistency check of readout value (not negative, maximum rate)
##### 2.3.0 (2019-10-05)
* Load default configuration, if none is present (beneficial for Docker-Version with mounted config and log directories)
* Parameter "simple" to reduche output to a single value
##### 2.2.0 (2019-09-18)
* Update neural network for readout analog meter
* storage and usage of last full readouts to substitute "NaN" values in digital counters
##### 2.1.0 Integration of direct request to Image provider
* The image soucre can be specified within in the config.ini. If no "url"-Parameter is given, the image will be pulled directly from the given source, including logging.
* Update of the name convention of the log-files from Unix-Timestamp to Human-Readable time coding
##### 2.0.0 Implementation in Python
* Usage of improved analog detection (sinus and cosinus coding)
* New folder structure to support easy implementation in Docker-ContainerUp
* **Attention:** adaption of INI-File needed, as Python is handling this slightly different 
##### 1.x.y Initial Version within NodeJS
* Implemenation of neural networks with version 1.x (analog and digital)
neuron

code/wasseruhr.py
from http.server import HTTPServer, BaseHTTPRequestHandler
from urllib import parse
import lib.ZaehlerstandClass
import os
import subprocess

import socketserver

import gc

class SimpleHTTPRequestHandler(BaseHTTPRequestHandler):
    
    def do_GET(self):
        global wasserzaehler
        url_parse = parse.urlparse(self.path)
        query_parse = parse.parse_qs(url_parse.query)

        if 'reload' in url_parse.path:
            self.send_response(200)
            self.send_header('Content-type', 'text/html')
            self.end_headers()
            result = 'Konfiguration wird neu geladen'
            self.wfile.write(bytes(result, 'UTF-8'))             
            del wasserzaehler
            gc.collect()
            wasserzaehler = lib.ZaehlerstandClass.Zaehlerstand()
            return

        if ('version' in url_parse.path) or ('ROI' in url_parse.path):
            result = "Version 7.1.1 (2020-05-24)"
            self.send_response(200)
            self.send_header('Content-type', 'text/html')
            self.end_headers()
            self.wfile.write(bytes(result, 'UTF-8'))
            return            

        GlobalError = wasserzaehler.CheckError()
        if GlobalError is not None:
            self.send_response(200)
            self.send_header('Content-type', 'text/html')
            self.end_headers()
            self.wfile.write(bytes(GlobalError, 'UTF-8'))

            return
           

        if "/image_tmp/" in url_parse.path:
            self.send_response(200)
            size = str(os.stat('.'+self.path).st_size)
            self.send_header('Content-type', 'image/jpg')
            self.send_header('Content-length', size)
            self.end_headers()
            with open('.'+self.path, 'rb') as file: 
                self.wfile.write(file.read()) # Read the file and send the contents
            return

        url = ''
        if 'url' in query_parse:
            url = query_parse['url'][0]

        simple = True
        if ('&full' in self.path) or ('?full' in self.path):
            simple = False

        single = False
        if ('&single' in self.path) or ('?single' in self.path):
            single = True

        usePrevalue = False
        if ('&usePreValue' in self.path) or ('?usePreValue' in self.path) or ('&usePrevalue' in self.path) or ('?usePrevalue' in self.path):
            usePrevalue = True

        value = ''
        if 'value' in query_parse:
            value = query_parse['value'][0]

        if ('crash' in url_parse.path):
            a = 1
            b = 0
            c = a / b
            return

        if ('crash' in url_parse.path):
            result = "Crash in a second"
            self.send_response(200)
            self.send_header('Content-type', 'text/html')
            self.end_headers()
            self.wfile.write(bytes(result, 'UTF-8'))
            print('Crash with division by zero!')
            a = 1
            b = 0
            c = a/b
            return

        if ('roi' in url_parse.path) or ('ROI' in url_parse.path):
            result = wasserzaehler.getROI(url)
            self.send_response(200)
            self.send_header('Content-type', 'text/html')
            self.end_headers()
            self.wfile.write(bytes(result, 'UTF-8'))
            return

        if 'setPreValue' in url_parse.path:
            result = wasserzaehler.setPreValue(value)
            self.send_response(200)
            self.send_header('Content-type', 'text/html')
            self.end_headers()
            self.wfile.write(bytes(result, 'UTF-8'))
            return

        if 'wasserzaehler.json' in url_parse.path:
            result = wasserzaehler.getZaehlerstandJSON(url, simple, usePrevalue, single)
            self.send_response(200)
            self.send_header('Content-type', 'application/json')
            self.end_headers()
            self.wfile.write(bytes(result, 'UTF-8'))
            return

        if 'wasserzaehler' in url_parse.path:
            result = wasserzaehler.getZaehlerstand(url, simple, usePrevalue, single)
            self.send_response(200)
            self.send_header('Content-type', 'text/html')
            self.end_headers()
            self.wfile.write(bytes(result, 'UTF-8'))
            return

if __name__ == '__main__':

    wasserzaehler = lib.ZaehlerstandClass.Zaehlerstand()

    PORT = 3000
    with socketserver.TCPServer(("", PORT), SimpleHTTPRequestHandler) as httpd:
        print("Wasserzaehler is serving at port", PORT)
        httpd.serve_forever()
